### ISSUE TITLE
Pinch point - very slow harmonisation

### ISSUE THREAD

**Issue by nicksunderland** [MRCIEU/TwoSampleMR#455]
In the harmonisation function, the way of finding the within SNP index (i.e. order of SNP appearance in the data) is extremely slow. The main issue is this line: https://github.com/MRCIEU/TwoSampleMR/blame/592ebe05538558b330c39ddeda0d11b1313ad819/R/harmonise.R#L553

This can be computed in ~0.09% of the time it currently takes.

e.g. 
```
library(microbenchmark)
library(ggplot2)
library(data.table)
n = 1e3
dat_df <- data.frame("SNP" = rep(LETTERS[1:10], times=n))
dat_dt <- data.table("SNP" = rep(LETTERS[1:10], times=n))
mbm <- microbenchmark("current"  = { SNP_index <- sapply(1:length(dat_df$SNP), function(i) sum(dat_df$SNP[1:i]==dat_df$SNP[i])) },
                      "proposed" = { dat_dt[, SNP_index := 1:.N, by="SNP"] },
                      times = 100L)
autoplot(mbm)
```

**Comment by nicksunderland** [comment]
doesn't have all the options of 2SMR, but is fast [source code](https://github.com/nicksunderland/genepi.utils/blob/main/R/harmonise.R). Speed comparison: [Vignette](https://nicksunderland.github.io/genepi.utils/articles/harmonise.html)

**Comment by remlapmot** [comment]
Great spot. This is very helpful, thanks indeed.

I have included it in #456 with acknowledgement to you (and co-authored commit).

**Comment by remlapmot** [comment]
Closed by #456
